File: src/main/java/org/elmlang/intellijplugin/features/structure/ElmUnionTypeTreeElement.java
package org.elmlang.intellijplugin.features.structure;

import com.intellij.ide.structureView.StructureViewTreeElement;
import com.intellij.ide.structureView.impl.common.PsiTreeElementBase;
import org.elmlang.intellijplugin.ElmIcons;
import org.elmlang.intellijplugin.psi.ElmTypeDeclaration;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import java.util.Collection;
import java.util.Collections;
import java.util.Optional;

class ElmUnionTypeTreeElement extends PsiTreeElementBase<ElmTypeDeclaration> {

    ElmUnionTypeTreeElement(ElmTypeDeclaration psiElement) {
        super(psiElement);
    }

    @Override
    public Icon getIcon(boolean open) {
        return ElmIcons.UNION_TYPE;
    }

    @Nullable
    @Override
    public String getPresentableText() {
        return Optional.ofNullable(getElement())
                .map(e -> e.getUpperCaseId().getName())
                .orElse(null);
    }

    @NotNull
    @Override
    public Collection<StructureViewTreeElement> getChildrenBase() {
        return Collections.emptyList();
    }
}

File: src/main/java/org/elmlang/intellijplugin/features/structure/ElmUnionTypeTreeElement.kt
package org.elmlang.intellijplugin.features.structure

import com.intellij.ide.structureView.StructureViewTreeElement
import com.intellij.ide.structureView.impl.common.PsiTreeElementBase
import com.intellij.psi.presentation.java.SymbolPresentationUtil
import org.elmlang.intellijplugin.ElmIcons
import org.elmlang.intellijplugin.psi.ElmTypeDeclaration

class ElmUnionTypeTreeElement(element: ElmTypeDeclaration): PsiTreeElementBase<ElmTypeDeclaration>(element) {

    override fun getIcon(open: Boolean) =
            ElmIcons.UNION_TYPE

    override fun getPresentableText(): String? {
        val decl = element
        return when (decl) {
            is ElmTypeDeclaration ->
                decl.upperCaseId.name
            else ->
                decl?.let { SymbolPresentationUtil.getSymbolPresentableText(it) }
        }
    }

    override fun getChildrenBase(): Collection<StructureViewTreeElement> =
            emptyList()
}
